enhancements/api-review/custom-tags-aws.md

```diff
@@ -79,65 +101,373 @@ The userTags field is intended to be set at install time and is considered immut
+
+   Action = Update existing tag with value from `.spec.platformSpec.aws.resourceTags`.
+
+   Final tag set to AWS resource = `key_infra1 = value_infra1`
+
+   Event action = An event is generated to notify user about the action status (success/failure) to update tags for the AWS resource.
+
+2) User deletes the user-defined tag from `.spec.platformSpec.aws.resourceTags`
+   The user-defined tag which is removed from spec, will not be reconciled or managed by operators.
+
+   User can update user-defined tag key:value using external tool. The user-defined tag will not be overwritten.
+
+   For example,
+
+   Existing tag for AWS resource = `key_infra1 = value1`
+
+   New tag request = `.spec.platformSpec.aws.resourceTags` has no user-defined tag with key `key_infra1`
+
+   Action = No change in tags for the AWS resource.
+
+   Final tag set to AWS resource = `key_infra1 = value1`
+
+   Event action = No event
+
+3) User sets user-defined tag created using external tools to empty string for deleting for AWS resource.
+
+   There is no validation check involved for creator tool of user-defined tag. Any user-defined tag listed by user in `.spec.platformSpec.aws.resourceTags` is considered for create/update/delete accordingly.
+
+4) Any user-defined tag set using `.spec.platformSpec.aws.resourceTags` in `Infrastructure.config.openshift.io/v1` type will affect all AWS resources managed.
+
+5) User-defined tags are not synced from `.spec.platformSpec.aws.resourceTags` to `.status.platformStatus.aws.resourceTags` for the following reasons.
+
+- User-defined tags in `.spec.platformSpec.aws.resourceTags` can be "removed without delete", "delete", "update".
+  `.spec.platformSpec.aws.resourceTags` to `.status.platformStatus.aws.resourceTags` sync is not required,
+  as there will be no versioning of the user-defined tags required to override the user-defined tags in Infrastructure CR.
+  Instead, the user-defined tags (if supported in resource operator spec field, e.g: machine) can override the user-defined tags in Infrastructure CR.
+
+- As `.spec.platformSpec.aws.resourceTags` has the actual values. Sync to `.status.platformStatus.aws.resourceTags` was required in earlier design to identify if the
+  tag on the resource was created using Infrastructure CR or external tool. Identifying the creator tool was done to restrict user from editing the same user-tag kv pair using multiple tools.
+  This inherently poses many scenarios of conflict which will result in user-defined tag kv pair being inconsistent across cluster when applied using Infrastructure CR. Hence, user will be confused which tool to be used to update tag.
+
+6) User applies user-defined tag using `.spec.platformSpec.aws.resourceTags`. Later, user modifies the user-defined tag for AWS resource using external tool.
+   In this case, the desired value is not set immediately for AWS resource by the owning operator. There is eventual consistency maintained by the owning operator.
```

Comment:
This is in strict opposition to the behavior that is described earlier in the EP. If there is eventual consistency, then without any further user action, the tag would eventually be updated on the AWS resource. The EP has stated previously that the tag on the AWS resource will not be updated until the user-defined tags are modified.

Reply:
As per the general desired value reconciliation, it is better to reconcile any updates. In the previous, version it was discarded taking into consideration for rate limiting and traffic generated. After discussion with @JoelSpeed , concluded that it is not much of a concern on rate limiters for the APIs and is implementation specific e.g: avoiding  adding any new API calls in existing reconcile loops.

Reply:
> The EP has stated previously that the tag on the AWS resource will not be updated until the user-defined tags are modified.

This seems to be against kube-patterns doesn't it? We have a desired state defined within the cluster, if anyone modifies the actual state, a controller puts it back. 

If you want to have the controllers only act when there's an update to the tags defined on the infra resource, then each controller will need to keep track of which tags it previously applied, then calculate a diff, and ensure the ordering is consistent to avoid spurious diffs. This seems overly complex and isn't something we would normally do in OpenShift, are we really sure we want to introduce this now?

Reply:
I think we have two choices here.

1) We guarantee that a tag on an AWS resource will be restored within a certain period of time after the resource is updated externally.
2) Or, we guarantee that a tag on an AWS resource will be restored after either the `resourceTags` field on the Infrastructure resource is updated or the tags on the kube resource is updated. The controller may restore the tag on the AWS resource prior to that. But there is no guarantee that the controller will restore the tag unless there is a change to a resource in the cluster that prompts the controller to check whether the tag needs to e updated.

I think that option (1) is the better user experience, but I understand if there are some controllers that may have limitations that preclude that as a viable option. With either option that we choose, the EP needs to be consistent.

Reply:
I certainly think option 1 is the preferred choice as it has better UX and is, IMO, easier to implement. Certainly for Machine API, we reconcile the status of our AWS resources at most every 10 minutes, lets check when the other controllers do so and see if we can put an upper time bound on this.

Is there a value we are happy with for the upper time bound? 10/30/60 mins?

Reply:
I care less about stipulating a specific upper time bound and more about ensuring that the EP is consistent with expectations. I am fine with leaving this part as saying that we ensure eventual consistency, so long as the previous part of the EP claiming that the tags on the AWS resource will only be updated in response to a spec change is updated to match this part.